speechbrain/lobes/models/wav2vec.py

```diff
@@ -0,0 +1,610 @@
+            num_vars = x["num_vars"]
+            temp = x["temp"]
+            diversity_loss = (
+                        num_vars - prob_perplex
+                    ) / num_vars
+            results.update(
+                {
+                    "diversity_loss": diversity_loss,
+                    "code_perplex": code_perplex,
+                    "prob_perplex": prob_perplex,
+                    "num_vars": num_vars,
+                    "temp": temp,
+                }
+            )
+        results["targets"] = targets
+        results["embeddings"] = feats
+        results["latent_l2"] = latent_l2
+        return results
+
+    def freeze(self):
+        for param in self.parameters():
+            param.requires_grad = False
+
+    def unfreeze(self):
+        for param in self.parameters():
+            param.requires_grad = True
+
+    def freeze_latent_extractor(self):
+        for param in self.latent_extractor.parameters():
+            param.requires_grad = False
+
+    def unfreeze_latent_extractor(self):
+        for param in self.latent_extractor.parameters():
+            param.requires_grad = True
+
+    @contextmanager
+    def no_sync(self):
+        """Dummy method to avoid if/else statements depending on if DDP used."""
+        yield
+
+
+class ContrastiveLoss(nn.Module):
+    """wav2vec specific contrastive loss
+    """
+
+    def __init__(self, logit_temp):
+        super().__init__()
+        self.logit_temp = logit_temp
+
+    def forward(self, x, y, negs):
+        """
+        Arguments
+        ----------
+        x : torch.Tensor
+            Encoded embeddings with shape (B, T, C).
+        y : torch.Tensor
+            Feature extractor target embeddings with shape (B, T, C).
+        negs : torch.Tensor
+            Negative embeddings from feature extractor with shape (N, B, T, C)
+        where N is number of negatives.
+        """
+        neg_is_pos = (y == negs).all(-1)
+        y = y.unsqueeze(0)
+        targets = torch.cat([y, negs], dim=0)
+        logits = torch.cosine_similarity(
+            x.float(), targets.float(), dim=-1
+        ).type_as(x)
+
+        if neg_is_pos.any():
+            logits[1:][neg_is_pos] = float("-inf")
+        logits = logits.transpose(0, 2).reshape(-1, logits.size(0))
+
+        targets = torch.zeros(
+            (logits.size(0)), dtype=torch.long, device=logits.device
+        )
+        loss = F.cross_entropy(
+            logits / self.logit_temp, targets, reduction="sum"
+        )
+        accuracy = torch.sum(logits.argmax(-1) == 0) / (
+            logits.numel() / logits.size(-1)
+        )
+        return loss, accuracy
+
+
+def make_padding_mask(inp_shape, abs_lens):
+    """False when to be used, following speechbrain convention """
+    padding_mask = (
+        torch.arange(inp_shape[1], device=abs_lens.device)[None, :] >= abs_lens[:, None]
+    )
+    return padding_mask
+
+
+def compute_mask(shape, padding_mask, mask_prob, mask_length):
```

Comment:
Could this also be a more general util function ? I don't think it is specific to w2v2. ?